File: android/src/main/kotlin/com/raftdb/RaftCollection.kt
package com.raftdb

import kotlinx.coroutines.flow.Flow
import kotlinx.coroutines.flow.map

/**
 * A typed, collection-scoped wrapper around [RaftDb].
 *
 * All keys are automatically prefixed with `<collection>:` so that
 * multiple collections can coexist in the same database without
 * key collisions.
 *
 * @param T the document type stored in this collection.
 * @property db the underlying [RaftDb] instance.
 * @property name the collection name used as key prefix.
 * @property serialize converts a [T] to bytes for storage.
 * @property deserialize converts stored bytes back to [T].
 */
class RaftCollection<T>(
    private val db: RaftDb,
    val name: String,
    private val serialize: (T) -> ByteArray,
    private val deserialize: (ByteArray) -> T,
) {

    private val prefix: ByteArray = "$name:".toByteArray(Charsets.UTF_8)

    /**
     * Insert or update a document by [id].
     *
     * @throws RaftError on native failure.
     */
    suspend fun put(id: String, document: T) {
        db.put(scopedKey(id), serialize(document))
    }

    /**
     * Retrieve a document by [id].
     *
     * @return the deserialized document, or `null` if not found.
     * @throws RaftError on native failure (other than not-found).
     */
    suspend fun get(id: String): T? {
        val bytes = db.get(scopedKey(id)) ?: return null
        return deserialize(bytes)
    }

    /**
     * Delete a document by [id]. Deleting a non-existent id is not an error.
     *
     * @throws RaftError on native failure.
     */
    suspend fun delete(id: String) {
        db.delete(scopedKey(id))
    }

    /**
     * Observe changes to a specific document by raw-KV [id] (legacy).
     *
     * Emits the current value once. The underlying raw-KV observer is
     * a stub; use [observe] (no-arg) for proper typed-FFI notifications.
     */
    fun observe(id: String): Flow<T?> =
        db.observe(scopedKey(id)).map { result ->
            result.value?.let(deserialize)
        }

    /**
     * Observe all changes in this collection via the raw-KV stub (legacy).
     */
    fun observeAll(): Flow<QueryResult> = db.observe(prefix)

    private fun scopedKey(id: String): ByteArray =
        "$name:$id".toByteArray(Charsets.UTF_8)

    // ── Typed-FFI surface (separate storage namespace from String-id) ──

    /**
     * Insert a document into the typed collection store, letting the
     * engine assign a fresh `Long` (uint64) id. Returns the assigned id.
     */
    suspend fun putAuto(document: T): Long =
        db.collectionPutAuto(name, serialize(document))

    /**
     * Insert or update a document at the given `Long` id. The
     * serialized document's `id` field must equal [docId].
     */
    suspend fun putById(docId: Long, document: T) {
        db.collectionPut(name, serialize(document))
    }

    /**
     * Retrieve a document by `Long` id from the typed namespace.
     * Returns `null` if not found.
     */
    suspend fun getById(docId: Long): T? {
        val bytes = db.collectionGet(name, docId) ?: return null
        return deserialize(bytes)
    }

    /**
     * Delete a document by `Long` id. Not an error if the id does
     * not exist.
     */
    suspend fun deleteById(docId: Long) {
        db.collectionDelete(name, docId)
    }

    /**
     * Number of documents in this collection (typed namespace).
     */
    suspend fun count(): Long = db.collectionCount(name)

    /**
     * All document ids in this collection (typed namespace),
     * sorted ascending.
     */
    suspend fun listIds(): LongArray = db.collectionListIds(name)

    /**
     * Observe every insert / update / delete on this collection
     * (typed namespace). Emits a [MutationEvent] per change.
     */
    fun observe(): Flow<MutationEvent> = db.observeCollection(name)

    /**
     * Observe a live query against this collection. Emits a
     * [QueryDiff] immediately with the initial snapshot, then again
     * every time the result set changes.
     */
    fun liveQuery(queryJson: ByteArray): Flow<QueryDiff> = db.observeQuery(queryJson)
}

File: swift/Sources/RaftDB/RaftCollection.swift
import Foundation

/// A typed wrapper around ``RaftDB`` with two complementary surfaces:
///
/// 1. **String-id raw-KV facade** — `put(id:document:)`, `get(id:)`,
///    `delete(id:)`. Documents are stored under `<name>:<id>` raw KV keys.
///    Useful when you control the id space (e.g. UUIDs, slugs).
///
/// 2. **Typed-FFI facade** — `putAuto(_:)`, `put(docId:document:)`,
///    `get(docId:)`, `delete(docId:)`, `count()`, `listIds()`,
///    `observe()`. Backed by Raft's typed collection store; document
///    ids are 64-bit unsigned integers assigned by the engine (or chosen
///    by the caller).
///
/// The two facades address **different storage namespaces**. Putting a
/// document via the String-id facade and reading it via the typed
/// facade will not work, and vice versa. Pick one per collection.
///
/// ```swift
/// struct User: Codable, Sendable { let id: UInt64; let name: String }
///
/// let db = try await RaftDB.open(path: path)
/// let users = RaftCollection<User>(db: db, name: "users")
///
/// // Typed facade
/// let id = try await users.putAuto(User(id: 0, name: "Alice"))
/// let alice = try await users.get(docId: id)
/// ```
public final class RaftCollection<T: Codable>: Sendable where T: Sendable {

    private let db: RaftDB
    public let name: String
    private let encoder: JSONEncoder
    private let decoder: JSONDecoder

    public init(
        db: RaftDB,
        name: String,
        encoder: JSONEncoder = JSONEncoder(),
        decoder: JSONDecoder = JSONDecoder()
    ) {
        self.db = db
        self.name = name
        self.encoder = encoder
        self.decoder = decoder
    }

    // MARK: - String-id raw-KV facade

    /// Insert or update a document under the String key `<name>:<id>`.
    public func put(id: String, document: T) async throws {
        let data = try encoder.encode(document)
        try await db.put(key: scopedKey(id), value: data)
    }

    /// Retrieve a document by String id from the raw-KV namespace.
    /// Returns `nil` if not found.
    public func get(id: String) async throws -> T? {
        guard let data = try await db.get(key: scopedKey(id)) else {
            return nil
        }
        return try decoder.decode(T.self, from: data)
    }

    /// Delete a document by String id from the raw-KV namespace. Not an
    /// error if the id does not exist.
    public func delete(id: String) async throws {
        try await db.delete(key: scopedKey(id))
    }

    // MARK: - Typed-FFI facade

    /// Insert a document into the typed collection store, letting the
    /// engine assign a fresh `UInt64` id. Returns the assigned id.
    public func putAuto(_ document: T) async throws -> UInt64 {
        let data = try encoder.encode(document)
        return try await db.collectionPutAuto(name, document: data)
    }

    /// Insert or update a document at the given `UInt64` id. The
    /// document's `id` field must equal `docId`.
    public func put(docId: UInt64, document: T) async throws {
        let data = try encoder.encode(document)
        try await db.collectionPut(name, document: data)
        _ = docId  // declared in signature to match call ergonomics with get/delete
    }

    /// Retrieve a document by `UInt64` id from the typed collection
    /// store. Returns `nil` if not found.
    public func get(docId: UInt64) async throws -> T? {
        guard let data = try await db.collectionGet(name, id: docId) else {
            return nil
        }
        return try decoder.decode(T.self, from: data)
    }

    /// Delete a document by `UInt64` id. Not an error if the id does
    /// not exist.
    public func delete(docId: UInt64) async throws {
        try await db.collectionDelete(name, id: docId)
    }

    /// Number of documents currently in this collection (typed
    /// namespace only).
    public func count() async throws -> Int {
        try await db.collectionCount(name)
    }

    /// All document ids currently in this collection (typed namespace),
    /// sorted ascending.
    public func listIds() async throws -> [UInt64] {
        try await db.collectionListIds(name)
    }

    // MARK: - Observation

    /// Observe every insert / update / delete on this collection
    /// (typed namespace). Yields a ``MutationEvent`` per change.
    public func observe() -> AsyncStream<MutationEvent> {
        db.observe(collection: name)
    }

    /// Observe a live query whose JSON encoding is `queryJson`. Yields
    /// a ``QueryDiff`` immediately with the initial snapshot and again
    /// every time the result set changes.
    public func liveQuery(_ queryJson: Data) -> AsyncStream<QueryDiff> {
        db.observeQuery(queryJson)
    }

    // MARK: - Internal

    private func scopedKey(_ id: String) -> Data {
        Data("\(name):\(id)".utf8)
    }
}
